### PR TITLE
Built-in Nougat support

### PR DESCRIPTION
**Pull request by sowelipililimute**
This PR allows using NuGet to simplify the management of plugins and their dependencies.

grafik

### CODE REVIEW

**Comment by Arthri** [comment]
IMO, the plugin manager code should be split into a separate executable/project

**Comment by sowelipililimute** [comment]
the plugin manager code now lives in its own project

File: TShockLauncher/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/packages
```
Comment (hakusaro) [inline TShockLauncher/.gitignore:1]: Can you explain why you added this to gitignore?

File: TShockLauncher/TShockLauncher.csproj [hunk 2]
```diff
@@ -50,27 +51,22 @@
 	</ItemGroup>
 
 	<Target Name="CheckMsgfmtCallable">
-		<Exec Command="msgfmt --help > NUL" IgnoreExitCode="True" Condition=" '$(OS)' == 'Windows_NT' ">
+		<Exec Command="msgfmt --help &gt; NUL" IgnoreExitCode="True" Condition=" '$(OS)' == 'Windows_NT' ">
```
Comment (hakusaro) [inline TShockLauncher/TShockLauncher.csproj]: Can you explain why this changed?

File: TShockLauncher/TShockLauncher.csproj [hunk 3]
```diff
@@ -50,27 +51,22 @@
...
 			<Output TaskParameter="ExitCode" PropertyName="MsgfmtExitCode" />
 		</Exec>
-		<Exec Command="msgfmt --help 2>/dev/null >/dev/null" IgnoreExitCode="True" Condition=" '$(OS)' != 'Windows_NT' ">
+		<Exec Command="msgfmt --help 2&gt;/dev/null &gt;/dev/null" IgnoreExitCode="True" Condition=" '$(OS)' != 'Windows_NT' ">
```
Comment (hakusaro) [inline TShockLauncher/TShockLauncher.csproj]: Can you explain why this changed? I'm not sure I see the necessity to this.

File: TShockLauncher/TShockLauncher.csproj [hunk 4]
```diff
@@ -50,27 +51,22 @@
...
 	</Target>
 
 	<!-- The condition for a Target can't come from the values of another target, so instead we have to put the same condition on all three of the items inside. -->
-	<Target
-		Name="GenerateMOFiles"
-		DependsOnTargets="CheckMsgfmtCallable"
-		AfterTargets="PostBuildEvent;Publish"
-		Inputs="..\i18n\**\*.po"
-		Outputs="$(OutDir)i18n\**\*.mo">
+	<Target Name="GenerateMOFiles" DependsOnTargets="CheckMsgfmtCallable" AfterTargets="PostBuildEvent;Publish" Inputs="..\i18n\**\*.po" Outputs="$(OutDir)i18n\**\*.mo">
```
Comment (hakusaro) [inline TShockLauncher/TShockLauncher.csproj]: This change seems unrelated entirely to the other changes in this PR -- and notably, makes it harder to read.

File: TShockPluginManager/NugetCLI.cs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+using NuGet.Versioning;
+
+namespace TShockPluginManager
+{
+	public static class NugetCLI
+	{
+		static public async Task<int> Main(List<string> args)
+		{
+			RootCommand root = new RootCommand(
+				description: "Manage plugins and their requirements"
```
Comment (hakusaro) [inline TShockPluginManager/NugetCLI.cs]: At a really high level, this is true of all the strings in this change set: I think that anything added to TShock, going forward, should be internationalized. It seems like a server operator could use most of TShock, just not the nuget-enabled TShock plugin system in their own language. But this is going to be relatively important. Can you add this support?

File: TShockPluginManager/NugetCLI.cs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+			SyncManifest manifest;
+			try
+			{
+				string txt = await File.ReadAllTextAsync("packages.json");
+				manifest = JsonSerializer.Deserialize<SyncManifest>(txt, opts)!;
+			}
+			catch (System.IO.FileNotFoundException)
+			{
+				System.Console.WriteLine("You're trying to sync, but you don't have a packages.json file.");
+				System.Console.WriteLine("Without a list of plugins to install, I can't install any.");
```
Comment (hakusaro) [inline TShockPluginManager/NugetCLI.cs]: In general, TShock conventionally uses third person or indirect speech / tone when speaking to users. Can you adjust the language on these strings to be more cohesive with the rest of the project?

```suggestion
				System.Console.WriteLine("Unable to install plugins without a list of plugins to install. Please supply a packages.json file to continue.");
```

To be as precise as possible: we try to avoid having the software speak as if it is a human or otherwise an intelligent agent. "Is it okay to connect to the internet?" vs "May I connect to the internet?" (in particular, I think this doesn't flow very well into other languages -- but the primary hangup is that it's tone that doesn't really exist in the project elsewhere.)

File: TShockPluginManager/NugetCLI.cs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+				return;
+			}
+			catch (System.Text.Json.JsonException e)
+			{
+				System.Console.WriteLine("I had an issue reading the packages.json, because...");
```
Comment (hakusaro) [inline TShockPluginManager/NugetCLI.cs]:
```suggestion
				System.Console.WriteLine("An unexpected error occurred while reading packages.json. See exception details:");
```

File: TShockPluginManager/NugetCLI.cs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+			var initial = Console.ForegroundColor;
+			foreach (var item in manifest.GetPackageIdentities())
+			{
+				System.Console.ForegroundColor = System.ConsoleColor.Green;
+				System.Console.Write(item.Id);
+				System.Console.ForegroundColor = System.ConsoleColor.Black;
+				System.Console.Write($" [{item.Version}]\n");
+			}
+			Console.ForegroundColor = initial;
+			System.Console.WriteLine("These are the packages you asked me to install.");
```
Comment (hakusaro) [inline TShockPluginManager/NugetCLI.cs]:
```suggestion
			System.Console.WriteLine("These packages are pending installation.");
```

File: TShockPluginManager/NugetCLI.cs [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+			System.Console.WriteLine("May I connect to the internet to figure out what I need to download?");
```
Comment (hakusaro) [inline TShockPluginManager/NugetCLI.cs]:
```suggestion
			System.Console.WriteLine("Connect to the internet to resolve packages?");
```

File: TShockPluginManager/NugetCLI.cs [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+			System.Console.WriteLine("=== Dependencies ===");
+			foreach (var item in nonUserPackages)
+				DumpOne(item, builtins);
+
+			System.Console.WriteLine("Can I download and install the given packages?");
+			System.Console.WriteLine("Make sure that you trust the plugins you're installing.");
+			System.Console.WriteLine("If you want to know which plugins need which dependencies, press E.");
+
+			bool ok = false;
+			do
```
Comment (hakusaro) [inline TShockPluginManager/NugetCLI.cs:145]: I can't really speak to other people who wrote C# more than I do, but do people actually use do-while loops in practice? I really don't like them.

File: TShockPluginManager/NugetCLI.cs [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+					new NuGetVersionConverter()
+				}
+			};
+			SyncManifest manifest;
+			try
+			{
+				string txt = await File.ReadAllTextAsync("packages.json");
+				manifest = JsonSerializer.Deserialize<SyncManifest>(txt, opts)!;
+			}
+			catch (System.IO.FileNotFoundException)
```
Comment (hakusaro) [inline TShockPluginManager/NugetCLI.cs:82]: We don't really have conventions around using directives in TShock, but I personally prefer that you throw something like `using System.IO;` in and then not fully-qualify the namespace throughout the project. The issue is that again, it just seems like a lot of work and it's extra verbosity that isn't needed, but that's just me.

File: TShockPluginManager/NugetCLI.cs [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+						{
+							DumpGraph(pkg, packagesToInstall, builtins, 0);
+						}
+						System.Console.WriteLine("Can I download and install the given packages?");
+						System.Console.WriteLine("If you'd like to see which plugins need which dependencies again, press E.");
+						break;
+				}
+			} while (!ok);
+
+			await nugetter.DownloadAndInstall(userRequests);
```
Comment (hakusaro) [inline TShockPluginManager/NugetCLI.cs:167]: I think I need to quit C# and TShock if it means that I have to deal with async/await in this language more.

File: TShockPluginManager/NugetCLI.cs [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+			var initial = Console.ForegroundColor;
+
+			var indent = new String('\t', level);
+			System.Console.Write(indent);
+			if (level != 0)
+			{
+				Console.ForegroundColor = ConsoleColor.Black;
+				System.Console.Write("needs ");
+			}
+			Console.ForegroundColor = ConsoleColor.Green;
```
Comment (hakusaro) [inline TShockPluginManager/NugetCLI.cs]: Can you provide a sample in a comment (e.g., ASCII in comment) that shows what this is supposed to look like?

File: docs/changelog.md [hunk 14]
```diff
@@ -75,6 +75,7 @@ Use past tense when adding new entries; sign your name off when you add or chang
 
 * Add ability for items given to players to be inserted directly into their inventory instead of spawned as an item drop (@pontaoski)
 * Added support of `-lang` and `-language` flags for our i18n system. (@KawaiiYuyu)
+* Add built-in package management capabilities for plugins
```
Comment (hakusaro) [inline docs/changelog.md]: Can you move this to the new upcoming changes section?
```suggestion
* Added built-in package management capabilities for plugins. (@pontaoski)
```

File: docs/packages.md [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+```
+./TShock.Server plugins
+```
+
+Under Windows (cmd.exe):
+```
+TShock.Server plugins
+```
+
+The documentation for the commands is included in the help functionality.
```
Comment (hakusaro) [inline docs/packages.md:17]: We're trying to provide as much functionality / documentation in docs files as possible, in order to make it easier to explain to people how things work. Can you provide command syntax, usages, and examples here, instead?

File: docs/packages.md [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+This file primarily exists to document the `packages.json`.
```
Comment (hakusaro) [inline docs/packages.md:19]: Can you explain this in more layman terms for people who have no idea what this is? What does it compare to in other languages?

More broadly, when are plugins loaded from this system? How manual is this? How does it update things?

File: TShockPluginManager/Nuget.cs [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+				}
+				else
+				{
+					pkgPath = null;
+					// die somehow
+				}
+			}
+			return (pkgPath, packageReader);
+		}
+		public async Task<IEnumerable<SourcePackageDependencyInfo>> GetPackagesToInstallFor(PackageIdentity[] userRequest)
```
Comment (hakusaro) [inline TShockPluginManager/Nuget.cs:242]: I think the preferred style is a blank line before/after all methods? Can you add these?

File: TShockPluginManager/Nuget.cs [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+			libItems = libItems.Where(x => x.TargetFramework.Equals(libnearest));
+
+			var frameworkItems = packageReader.GetFrameworkItems();
+			var fwnearest = FrameworkReducer.GetNearest(NuGetFramework, frameworkItems.Select(x => x.TargetFramework));
+			frameworkItems = frameworkItems.Where(x => x.TargetFramework.Equals(fwnearest));
+
+			return libItems.Concat(frameworkItems);
+		}
+
+		public async Task<(string, PackageReaderBase)> GetOrDownloadPackage(SourcePackageDependencyInfo pkg)
```
Comment (hakusaro) [inline TShockPluginManager/Nuget.cs:195]: In-general, we try to have XML documentation for all public methods and members. Can you add this?

File: TShockPluginManager/Nuget.cs [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+		public static bool IsSubPathOf(string subPath, string basePath) {
+			var rel = Path.GetRelativePath(basePath, subPath);
+			return rel != "."
+				&& rel != ".."
+				&& !rel.StartsWith("../")
+				&& !rel.StartsWith(@"..\")
+				&& !Path.IsPathRooted(rel);
+		}
+
+		public IEnumerable<FrameworkSpecificGroup> ItemsWeCareAbout(PackageReaderBase packageReader)
```
Comment (hakusaro) [inline TShockPluginManager/Nuget.cs]: Why do we care about them? Can you document this method more?

File: TShockPluginManager/Nuget.cs [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+		private ISettings Settings;
+		private NuGetPathContext PathContext;
+		private PackageSourceProvider PackageSourceProvider;
+		private SourceRepositoryProvider SourceRepositoryProvider;
+		private PackagePathResolver PackagePathResolver;
+		private PackageExtractionContext PackageExtractionContext;
+
+		public Nugetter()
+		{
+			FrameworkReducer  = new FrameworkReducer();
```
Comment (hakusaro) [inline TShockPluginManager/Nuget.cs]: Can you consistently add only one space between things in this block of code?

**Review by hakusaro — CHANGES_REQUESTED** [review]
Thank you for your work on this! I'm sure it will really really help in terms of getting new plugins installed for server operators in the future.

I have a few review comments. If I were to summarize my main qualms with this it's the following:

1. Internally inconsistent style. Even within only additions and only new files in this project, the code style is internally inconsistent, in terms of sheer syntactical formatting. In particular, inconsistent spacing and indentation between methods, odd spacing between individual characters, and other syntactic choices make this difficult to review.
2. Lack of support for i18n -- we just worked really hard on internationalizing TShock. There are a few edge cases, namely in TSAPI (which doesn't currently use TShock i18n stuff), but broadly, I'm happy with what we've done so far. That being said, it necessarily implies that new code should support the i18n system. If not, we're just adding more debt to pay off in the future.
3. General maintainability / future documentation needs to be improved, imho. The biggest thing to remember with TShock is that this project has been around for a long, long time. We have had maintainers come and go and entire systems become maintained by us that we have no idea how they work. For example, the entire database engine / system was written by a charismatic New Zealander who hasn't been around in...a few years. We've had to put up with this, upgrade it, maintain it, etc. for a long time. Systems like this necessarily enter the realm of "nice to have", and therefore, the maintainability requirements go up significantly. If SSC doesn't work, that's a core feature we _have_ to fix. If this doesn't work and nobody knows how to fix it, the chances of someone fixing it go down with the complexity and documentation overload. We used to have a really fancy automatic update system that would download and install and reboot TShock with updates installed. We removed that system because it was a lot of overhead/complexity to maintain and it wasn't very well-understood. This same fate will befall this system unless the code is rock solid, easily understood, and maintainable.

To elaborate more on point 3, let me recall a couple variable declaration blocks:

```csharp
		private FrameworkReducer FrameworkReducer;
		private NuGetFramework NuGetFramework;
		private ISettings Settings;
		private NuGetPathContext PathContext;
		private PackageSourceProvider PackageSourceProvider;
		private SourceRepositoryProvider SourceRepositoryProvider;
		private PackagePathResolver PackagePathResolver;
		private PackageExtractionContext PackageExtractionContext;
```

and

```csharp
			var nugetter = new Nugetter();
			var userRequests = manifest.GetPackageIdentities();
			var packagesToInstall = await nugetter.GetPackagesToInstallFor(manifest.GetPackageIdentities());
			var builtins = await nugetter.GetAllBuiltinDependencies();
			var userRoots = packagesToInstall.Where(x => userRequests.Any(y => x.Id == y.Id));
			var nonUserPackages = packagesToInstall.Where(x => !userRequests.Any(y => x.Id == y.Id));
```

In each of these instances, we "setup the world" ala COBOL or another language that supports this pattern. Then, the usages of the declared items happen significantly later and aren't super well-understood just by looking at them. I know broadly how package managers work, so I'm at an advantage here. I know what some of these words mean, and I can guess what the outputs are. But the overall control flow is opaque to a newcomer. `userRoots` _deserves_ a line or two about what a root element is and why it's important and what it's used for.

I would even support a broad architecture document that spells out, in plain English, what the high level process that this system uses is.

----

The bar for a lot of code in TShock is, unfortunately, variable. For code that isn't necessarily "core" code like what already exists, we have a relatively high bar: we want new stuff to be relatively well understood and well-patterned. This means things like documenting all public members, and even documenting private members when necessary. For existing code, like what you see in `Commands.cs`, there's a much lower barrier to entry, because we quite frankly know that we're dealing with an unwieldy beast (Terraria). The issue here is that this is entirely greenfield, and thus, there's a higher standard / barrier to entry to direct merge. I don't want to see this merged and abandoned, or merged, forgotten, and then never used. I don't want us to have to rip it all out in 5 years because it's too cumbersome to update for some reason.

On behalf of the other maintainers, I think I speak for all of them when I say that we want new code to be as clear, obvious, simple, maintainable, and consistent as possible.

Also, as an unrelated note: If you're familiar enough with git, we really prefer commits to be sensible too. For code changes, it becomes very difficult for us to find things if we have to look at the history if the commits are total chaos. For documentation and CI stuff that's really not important, but when using `git blame` commit messages ideally should have paragraphs explaining what changed and why and why it's beneficial, etc. I know that again, this is mostly greenfield, but I want to put-it-forward that it is something we care about. We just don't typically enforce these style items on some reviews (e.g., PRs that are clearly one-shots from people who can't use git), but if you can grace us with good git conventions, we'd really appreciate it too.

File: TShockPluginManager/Nuget.cs [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+			}
+		}
+		public async Task<IEnumerable<SourcePackageDependencyInfo>> GetAllBuiltinDependencies()
+		{
+			var knownBundles = new[] {
+				new PackageIdentity("GetText.NET", NuGetVersion.Parse("1.6.6")),
+				new PackageIdentity("OTAPI.Upcoming", NuGetVersion.Parse("3.1.8-alpha")),
+				new PackageIdentity("TSAPI", NuGetVersion.Parse("5.0.0-beta")),
+				new PackageIdentity("TShock", NuGetVersion.Parse("5.0.0-beta")),
+			};
```
Comment (QuiCM) [inline TShockPluginManager/Nuget.cs:117]: I'd like to avoid hard-coding versions if possible. Otherwise, when we update OTAPI/TSAPI/TShock we'll also have to keep version strings in sync here too

File: TShockLauncher/TShockLauncher.csproj (hunk 2 above)
Comment (sowelipililimute) [reply TShockLauncher/TShockLauncher.csproj]: dotnet tooling automatically poked it and i didn't notice

File: TShockPluginManager/NugetCLI.cs (hunk 10 above)
Comment (bartico6) [reply TShockPluginManager/NugetCLI.cs:145]: > but do people actually use do-while loops in practice? I really don't like them.  

Yes we do, for any task that requires you to do it once and then repeat it as long as a condition holds. The alternative would be to either duplicate code, or do a regular while loop with some `while(first_run || actual_condition)` gimmick.

**Comment by hakusaro** [comment]
@pontaoski if you're ready to have this reviewed can you de-conflict this branch and request re-reviews from me et al?

**Comment by hakusaro** [comment]
@pontaoski this will get re-reviewed, probably in 5.2 with the other PRs that are large-ish and unrelated to Terraria patch. Sorry, re-logic got us surprised again.

**Comment by hakusaro** [comment]
@pontaoski can you rebase?

**Comment by sowelipililimute** [comment]
rebased
